python_modules/dagster/dagster/core/definitions/dependency.py

```diff
@@ -258,10 +258,12 @@ def get_resource_requirements(
             solid_def = self.definition.ensure_solid_def()
             for requirement in solid_def.get_resource_requirements((cur_node_handle, asset_layer)):
                 # If requirement is a root input manager requirement, but the corresponding node has an upstream output, then ignore the requirement.
-                if isinstance(
-                    requirement, InputManagerRequirement
-                ) and outer_container.dependency_structure.has_deps(
-                    SolidInputHandle(self, solid_def.input_def_named(requirement.input_name))
+                if (
+                    isinstance(requirement, InputManagerRequirement)
+                    and outer_container.dependency_structure.has_deps(
+                        SolidInputHandle(self, solid_def.input_def_named(requirement.input_name))
+                    )
+                    and requirement.root_input
```

Comment:
i need a way to differentiate between inputs with a root input manager vs ones with a specific input manager. right now doing it by adding an attribute to InputManagerRequirement, but open to other ways of doing this

Reply:
This is fine IMO. Other option is another subclass but that seems excessive.